source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,107 @@ Example::
+structure::
+
+    {
+        isMaster: 1,
+        speculativeAuthenticate: {
+            authenticate: 1,
+            mechanism: "MONGODB-X509"
+        }
+    }
+
+When the mechanism is ``SCRAM-SHA-1`` or ``SCRAM-SHA-256``, `speculativeAuthenticate`
+has the following structure::
+
+    {
+        isMaster: 1,
+        speculativeAuthenticate: {
+            saslStart: 1,
+            mechanism: "<string>",   /* "SCRAM-SHA-1" or "SCRAM-SHA-256" */
+            db: "<authentication database>",
+            payload: BinData(...),
+            options: {
+                skipEmptyExchange: true         /* REQUIRED */
+            }
+        }
+    }  
+
+
+speculativeAuthenticate.mechanism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required.
+
+This value indicates which authentication mechanism to use with the credential.
+Currently, the only mechanism supported when ``speculativeAuthenticate.authenticate``
+is present is ``MONGODB-X509``.
+
+speculativeAuthenticate.db
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required.
+
+This value indicates the name of the database used for authentication. 
+The database used when ``speculativeAuthenticate.authenticate`` is present is
+``$external``. When ``speculativeAuthenticate.saslStart`` is present, the database
+name defaults to the database name provided in the connection string or ``admin``.
+
+speculativeAuthenticate.payload
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required when ``speculativeAuthenticate.saslStart`` is present.
+
+A sequence of bytes or a base64 encoded string (depending on input) to pass
+into the SASL library to transition the state machine.
+
+speculativeAuthenticate.options.skipEmptyExchange
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required when ``speculativeAuthenticate.saslStart`` is present.
+
+This value notifies newer servers to shorten the authentication conversation by
+one round trip. Older servers will ignore this field.
+
+If the ``isMaster`` command with a ``speculativeAuthenticate`` argument succeeds,
+the client should proceed with the next step of the exchange. If the ``isMaster``
+response does not include a ``speculativeAuthenticate`` reply and the ``ok`` field
+in the ``isMaster`` response is set to 1, drivers MUST authenticate using the standard
+authentication handshake.
+
+If an authentication mechanism is not provided either via connection string or code, but
+a credential is provided, drivers MUST use the SCRAM-SHA-256 mechanism for speculative
+authentication. If no authentication mechanism and no credential are provided, drivers
+MUST continue to authenticate using the standard authentication handshake.
+
+Older servers will ignore the ``speculativeAuthenticate`` argument. New servers will
+participate in the standard authentication conversation if this argument is missing.
+
+Drivers MUST set ``speculativeAuthenticate.saslStart`` if either of the authentication
```

Comment:
Given what I said above, this prose can be simplified a bit to just talk about use of `speculativeAuthenticate` in general rather than the specific fields within it.